vm: include vm context in the embedded snapshot

joyeecheung opened the pull request:
The first commit comes from https://github.com/nodejs/node/pull/44258

### vm: make ContextifyContext template context-independent

Instead of creating an object template for every ContextifyContext,
we now create one object template that can be reused by all
contexts. The native pointer can be obtained through an embdder
pointer field in the creation context of the receiver in the
interceptors, because the interceptors are only meant to be invoked
on the global object of the contextified contexts. This makes
the ContextifyContext template context-independent and therefore
snapshotable.

### vm: include vm context in the embedded snapshot

Include a minimally initialized contextify context in the embedded
snapshot. This paves the way for user-land vm context snapshots.

### vm: avoid unnecessary property getter interceptor calls

Access to the global object from within a vm context is intercepted
so it's slow, therefore we should try to avoid unnecessary access
to it during the initialization of vm contexts.

- Remove the Atomics.wake deletion as V8 now does not install it
  anymore.
- Move the Intl.v8BreakIterator deletion into the snapshot.
- Do not query the Object prototype if --disable-proto is not set.

This should speed up the creation of vm contexts by about ~12%.

Refs: https://github.com/nodejs/node/issues/44014
Refs: https://github.com/nodejs/node/issues/37476

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

mscdex commented:
s/embeded/embedded in commit message

joyeecheung commented on src/node_contextify.cc:
```diff
@@ -371,10 +401,26 @@ ContextifyContext* ContextifyContext::ContextFromContextifiedSandbox(
+  }
+  if (UNLIKELY(context->GetAlignedPointerFromEmbedderData(
+                   ContextEmbedderIndex::kContextTag) !=
+               Environment::kNodeContextTagPtr)) {
```
Do you want to open a separate PR for that? It would be better to rebase this one on top of that, instead of the entire https://github.com/nodejs/node/pull/44179

legendecas replied:
Yeah, I'll submit a separate PR for it.

joyeecheung commented:
Rebased on top of https://github.com/nodejs/node/pull/44258 and added a commit that makes use the snapshot to speed up the initialization of the vm contexts a bit.

goloveychuk commented:
I've found that 
```js
const vm = require('vm')

const context = vm.createContext();
console.log(context, vm.isContext(context))
const global2 = vm.runInContext(
    'this',
    context,
)
```
throws an error
```
TypeError: The "contextifiedObject" argument must be an vm.Context. Received an instance of Object
```
is it next thing to fix?

joyeecheung commented:
@goloveychuk My plan is to implement user-land vm context it in another PR (because that requires more refactoring to the lifetime management of the bindings), this only adds support for vm context in the embedded snapshot, which isn't that complex.

goloveychuk commented:
just fyi, with this PR and small changes inside `cjs/loader` I've managed cjs require to work. 
https://github.com/goloveychuk/node/commit/745bcef7b549b679667b9d26b83c00304ef58e0f (very dirty)
Good job!

joyeecheung commented:
@goloveychuk Thanks, that's good news. Though I've been thinking if we should only support loading user-modules at snapshot building time but disallowing that at runtime, and the other way around, but not both, or it can be quite a challenge to keep the module lookup in-sync unless the user uses some kind of docker container to build and deploy things. But that's probably a topic separate from this PR.

Marking it as blocked until https://github.com/nodejs/node/pull/44258 lands

joyeecheung commented:
Opened https://github.com/nodejs/node/issues/44277 and transferred the discussions there, in case this PR goes too off-topic. @goloveychuk please use that thread for discussions about user-land modules instead, thanks

legendecas commented on src/env.cc:
```diff
@@ -769,6 +778,8 @@ void Environment::InitializeMainContext(Local<Context> context,
                                         const EnvSerializeInfo* env_info) {
   context_.Reset(context->GetIsolate(), context);
   AssignToContext(context, ContextInfo(""));
+  context->SetAlignedPointerInEmbedderData(
```
Do we still need this? It seems like `ContextEmbedderIndex::kContextifyContext` should already be initialized to `nullptr` in `AssignToContext`.

joyeecheung replied:
Yeah it should no longer be necessary after AssignToContext initializes too. I've removed it.

legendecas commented on src/node_contextify.cc:
```diff
@@ -240,30 +256,39 @@ MaybeLocal<Context> ContextifyContext::CreateV8Context(
-  return scope.Escape(ctx);
+  // This should only be done after the initial initializations of the context
+  // global object is finished.
+  ctx->SetAlignedPointerInEmbedderData(ContextEmbedderIndex::kContextifyContext,
```
Do we still need this as `ContextEmbedderIndex::kContextifyContext` already initialized in `AssignToContext`?

legendecas replied:
nvm. I overlooked the second parameter.

joyeecheung commented:
Landed in e0191ca9fb63...7900f6540ea8

juanarbol commented:
This is not landing cleanly in v16.x
